### PR TITLE
[FIX] stock:  filter invalid input from domain

### PR DESCRIPTION
**Pull request by sami-odoo**
When the user tries to find the 'qty_available' of a product but gives
a string value instead of a float value in domain then the user will face error.

Steps to produce:
- Install `stock`.
- Create a record rule for model `product.template`. In the domain of record
  rule keep [('qty_available', '<', '10')] as a domain,
  you can give any integral value in the domain but that value should be
  enterred as strings (with quotes).
- Inventory > Products > Products

Traceback:
```
TypeError: '<' not supported between instances of 'str' and 'float'
  File "odoo/http.py", line 2134, in __call__
    response = request._serve_db()
  File "odoo/http.py", line 1710, in _serve_db
    return service_model.retrying(self._serve_ir_http, self.env)
  File "odoo/service/model.py", line 133, in retrying
    result = func()
  File "odoo/http.py", line 1737, in _serve_ir_http
    response = self.dispatcher.dispatch(rule.endpoint, args)
  File "odoo/http.py", line 1938, in dispatch
    result = self.request.registry['ir.http']._dispatch(endpoint)
  File "odoo/addons/base/models/ir_http.py", line 191, in _dispatch
    result = endpoint(**request.params)
  File "odoo/http.py", line 717, in route_wrapper
    result = endpoint(self, *args, **params_ok)
  File "addons/web/controllers/dataset.py", line 30, in call_kw
    return self._call_kw(model, method, args, kwargs)
  File "addons/web/controllers/dataset.py", line 26, in _call_kw
    return call_kw(request.env[model], method, args, kwargs)
  File "odoo/api.py", line 461, in call_kw
    result = _call_kw_multi(method, model, args, kwargs)
  File "odoo/api.py", line 448, in _call_kw_multi
    result = method(recs, *args, **kwargs)
  File "odoo/models.py", line 3202, in read
    self._origin.fetch(fields)
  File "odoo/models.py", line 3471, in fetch
    fetched = self._fetch_query(query, fields_to_fetch)
  File "odoo/models.py", line 3589, in _fetch_query
    field.read(fetched)
  File "odoo/fields.py", line 4729, in read
    wquery = comodel._where_calc(domain)
  File "odoo/models.py", line 4728, in _where_calc
    return expression.expression(domain, self).query
  File "odoo/osv/expression.py", line 777, in __init__
    self.parse()
  File "odoo/osv/expression.py", line 1125, in parse
    domain = field.determine_domain(model, operator, right)
  File "odoo/fields.py", line 1401, in determine_domain
    return determine(self.search, records, operator, value)
  File "odoo/fields.py", line 100, in determine
    return needle(*args)
  File "addons/stock/models/product.py", line 769, in _search_qty_available
    product_variant_query = self.env['product.product']._search(domain)
  File "home/odoo/src/enterprise/saas-16.4/stock_barcode/models/product_product.py", line 15, in _search
    return super()._search(domain, offset=offset, limit=limit, order=order, access_rights_uid=access_rights_uid)
  File "addons/product/models/product_product.py", line 423, in _search
    return super()._search(domain, offset, limit, order, access_rights_uid)
  File "odoo/models.py", line 4987, in _search
    query = self._where_calc(domain)
  File "odoo/models.py", line 4728, in _where_calc
    return expression.expression(domain, self).query
  File "odoo/osv/expression.py", line 777, in __init__
    self.parse()
  File "odoo/osv/expression.py", line 1125, in parse
    domain = field.determine_domain(model, operator, right)
  File "odoo/fields.py", line 1401, in determine_domain
    return determine(self.search, records, operator, value)
  File "odoo/fields.py", line 100, in determine
    return needle(*args)
  File "addons/stock/models/product.py", line 318, in _search_qty_available
    product_ids = self._search_qty_available_new(
  File "addons/stock/models/product.py", line 373, in _search_qty_available_new
    value < 0.0 and operator in ('>', '>=') or
```

After applying our commit, if the domain is Invalid then the user will face
`UserError`. There is a function named `_search_product_quantity`  which
initially checks if the domain is valid or not :
https://github.com/odoo/odoo/blob/5f1ad0c1e4c1e8ed41337ccd4c12e70c9c760c85/addons/stock/models/product.py#L342-L350
It will filter out all the invalid domain and clear search methods.
So now whenever `_search_qty_available` is called first the validity of domain
and operands of the expression will be checked.

sentry-4399815131

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by bit-odoo** [comment]
Hello @sami-odoo can you please check this for other models too? maybe need a generic fix here.

**Comment by sami-odoo** [comment]
Hello @bit-odoo sir,

> Hello @sami-odoo can you please check this for other models too?

I have checked this for other models but this case is arising from model 'product.template' and 'product.product' only
because 'qty_available' is a float field that has been created in these two models.

> maybe need a generic fix here.

Odoo's codebase handle's invalid domain but our domain is invalid in a manner that user has provided string value for a field that is float. For eg: `[('qty_available', '<', '10')]`  . Now, in domain `qty_available` is a float field but user has given 10 in the form of a string. So that is the reason user faced error similar to our traceback `'<' not supported between instances of 'str' and 'float'`.  It is a small corner case which can be handled by executing function `_search_product_quantity`

Also, i have updated commit message which highlights the specific case of our issue.

Thank you!

**Comment by sami-odoo** [comment]
Hello @odoo/sentry_reviewers ,
Please review it!

**Comment by HANNICHE-Walid** [comment]
looks good to me 
but can you also edit the error message 

right now it will display:
"Invalid domain right operand 10"
which doesn't convey that the error is actually it being a string not a number

**Comment by sami-odoo** [comment]
Hello @HANNICHE-Walid ,

I have updated the error message and made it more conveyable.

> it would also be better to check the validity of the domain during the rule creation rather than when using it

Pr: https://github.com/odoo/odoo/pull/122223

The above pr aims to check the validty of domain while rule creation

File: addons/stock/i18n/stock.pot [hunk 1]
```diff
@@ -2926,6 +2926,12 @@ msgstr ""
...
 
+#. module: stock
+#: code:addons/stock/models/product.py:0
+#, python-format
+msgid "Invalid domain right operand '%s'. It must be of type Integer/Float"
+msgstr ""
+
 #. module: stock
 #: code:addons/stock/models/product.py:0
 #, python-format
```
Comment (ticodoo) [inline addons/stock/i18n/stock.pot:2937]: For future reference, you need to update the existing .pot string that you are changing. But since your fix is wrong you do not need to change it in this case.

File: addons/stock/models/product.py [hunk 2]
```diff
@@ -288,13 +288,14 @@ def _search_qty_available(self, operator, value):
         # to use the quants, not the stock moves. Therefore, we bypass the usual
         # '_search_product_quantity' method and call '_search_qty_available_new' instead. This
         # allows better performances.
+        search_domain = self._search_product_quantity(operator, value, 'qty_available')
```
Comment (ticodoo) [inline addons/stock/models/product.py:291]: This change completely defeats the point of using `_search_qty_available_new` and avoiding using `_search_product_quantity` when it's not necessary (please read the code comment next time and try to understand what the code is doing before making your change). Please add new UserError checks to the `_search_qty_available_new` method instead

**Review by ticodoo — CHANGES_REQUESTED** [review]

File: addons/stock/i18n/stock.pot [hunk 3]
```diff
@@ -2921,15 +2921,15 @@ msgid "Invalid domain left operand %s"
 msgstr ""
 
 #. module: stock
-#: code:addons/stock/models/product.py:0
+#: code:addons/stock/models/product.py:0 code:addons/stock/models/product.py:0
```
Comment (ticodoo) [inline addons/stock/i18n/stock.pot:2924]: undo this change

File: addons/stock/i18n/stock.pot [hunk 4]
```diff
@@ -2921,15 +2921,15 @@ msgid "Invalid domain left operand %s"
...
 #, python-format
 msgid "Invalid domain operator %s"
 msgstr ""
 
 #. module: stock
-#: code:addons/stock/models/product.py:0
+#: code:addons/stock/models/product.py:0 code:addons/stock/models/product.py:0
```
Comment (ticodoo) [inline addons/stock/i18n/stock.pot:2930]: undo this change

File: addons/stock/i18n/stock.pot (hunk 3 above)
Comment (ticodoo) [reply addons/stock/i18n/stock.pot:2924]: I was wrong

**Comment by ticodoo** [comment]
@robodoo r+

**Comment by ticodoo** [comment]
@robodoo retry
